program/src/processor.rs

```diff
@@ -1641,6 +1642,84 @@ impl Processor {
+    /// instruction
+    pub fn process_unwrap_lamports(
+        program_id: &Pubkey,
+        accounts: &[AccountInfo],
+        amount: PodCOption<u64>,
+    ) -> ProgramResult {
+        let account_info_iter = &mut accounts.iter();
+        let source_account_info = next_account_info(account_info_iter)?;
+        let destination_account_info = next_account_info(account_info_iter)?;
+        let authority_info = next_account_info(account_info_iter)?;
+        let authority_info_data_len = authority_info.data_len();
+
+        let mut source_account_data = source_account_info.data.borrow_mut();
+        let source_account =
+            PodStateWithExtensionsMut::<PodAccount>::unpack(&mut source_account_data)?;
+
+        let (amount, remaining_amount) = match amount {
+            PodCOption {
+                option: PodCOption::<u64>::SOME,
+                value: amount,
+            } => (
+                amount,
+                Into::<u64>::into(source_account.base.amount)
+                    .checked_sub(amount)
+                    .ok_or(TokenError::InsufficientFunds)?,
+            ),
+            PodCOption {
+                option: PodCOption::<u64>::NONE,
+                value: _,
+            } => (source_account_info.lamports(), 0),
+            _ => return Err(ProgramError::InvalidInstructionData),
+        };
+
+        if !source_account.base.is_native() {
+            return Err(TokenError::NonNativeNotSupported.into());
+        }
+
+        let self_transfer = source_account_info.key == destination_account_info.key;
+        if let Ok(cpi_guard) = source_account.get_extension::<CpiGuard>() {
+            if *authority_info.key == source_account.base.owner
+                && cpi_guard.lock_cpi.into()
+                && in_cpi()
+            {
+                return Err(TokenError::CpiGuardTransferBlocked.into());
+            }
+        }
+
+        Self::validate_owner(
+            program_id,
+            &source_account.base.owner,
+            authority_info,
+            authority_info_data_len,
+            account_info_iter.as_slice(),
+        )?;
+
+        // Revisit this later to see if it's worth adding a check to reduce
+        // compute costs, ie:
+        // if self_transfer || amount == 0
+        check_program_account(source_account_info.owner)?;
+
+        if !self_transfer {            
+            let source_starting_lamports = source_account_info.lamports();
+            **source_account_info.lamports.borrow_mut() = source_starting_lamports
+            .checked_sub(amount)
+            .ok_or(TokenError::Overflow)?;
+            
+            let destination_starting_lamports = destination_account_info.lamports();
+            **destination_account_info.lamports.borrow_mut() = destination_starting_lamports
+            .checked_add(amount)
+            .ok_or(TokenError::Overflow)?;
+        
+            source_account.base.amount = remaining_amount.into();
```

Comment:
This is also present in the p-token implementation, and now I'm questioning it. We don't decrement the `source_account.amount` during a self-transfer, but it would definitely be clearer if we did. Meaning, if I unwrap from my wrapped SOL account into itself, I would imagine those lamports go from being part of `amount` to no longer being part of `amount`.

What do you think @febo ?

Reply:
Changed p-token implementation here [#113](https://github.com/solana-program/token/pull/113).